src/prototype_room_init.js

```diff
@@ -68,41 +68,76 @@ Room.prototype.initSetStorageAndPathStart = function() {
 
 Room.prototype.setFillerArea = function(storagePos, route) {
   const costMatrix = this.getMemoryCostMatrix();
+
   const fillerPos = storagePos.getBestNearPosition();
-  // TODO reset all values if the first doesn't fit
-  this.log('Testing fillerPos' + JSON.stringify(fillerPos));
-  this.deleteMemoryPath('pathStart-filler');
   this.memory.position.creep.filler = fillerPos;
   costMatrix.set(fillerPos.x, fillerPos.y, config.layout.creepAvoid);
   this.setMemoryCostMatrix(costMatrix);
+
   const pathFiller = this.getPath(route, 0, 'pathStart', 'filler', true);
-  for (const pos of pathFiller) {
-    this.increaseCostMatrixValue(costMatrix, pos, config.layout.pathAvoid);
+  this.setCostMatrixPath(costMatrix, pathFiller);
+  this.setMemoryCostMatrix(costMatrix);
+
+  const fillerNearPositions = Array.from(fillerPos.findNearPosition());
+  if (fillerNearPositions.length < 4) {
+    throw new Error(`Can't set layout for room ${this.name}. Not enough space for filler area`);
   }
+
+  const linkStoragePos = fillerNearPositions.shift();
+  this.memory.position.structure.link.unshift(linkStoragePos);
+  costMatrix.set(linkStoragePos.x, linkStoragePos.y, config.layout.structureAvoid);
   this.setMemoryCostMatrix(costMatrix);
-  const linkStoragePosIterator = fillerPos.findNearPosition();
-  for (const linkStoragePos of linkStoragePosIterator) {
-    this.memory.position.structure.link.unshift(linkStoragePos);
-    costMatrix.set(linkStoragePos.x, linkStoragePos.y, config.layout.structureAvoid);
-    this.setMemoryCostMatrix(costMatrix);
 
-    const powerSpawnPosIterator = fillerPos.findNearPosition();
-    for (const powerSpawnPos of powerSpawnPosIterator) {
-      this.memory.position.structure.powerSpawn.push(powerSpawnPos);
-      costMatrix.set(powerSpawnPos.x, powerSpawnPos.y, config.layout.structureAvoid);
-      this.setMemoryCostMatrix(costMatrix);
+  this.setPosition(STRUCTURE_POWER_SPAWN, fillerNearPositions.shift());
 
-      const towerPosIterator = fillerPos.findNearPosition();
-      for (const towerPos of towerPosIterator) {
-        this.memory.position.structure.tower.push(towerPos);
-        costMatrix.set(towerPos.x, towerPos.y, config.layout.structureAvoid);
-        this.setMemoryCostMatrix(costMatrix);
-        return;
-      }
-      this.memory.position.structure.powerSpawn.pop();
+  this.setPosition(STRUCTURE_TOWER, fillerNearPositions.shift());
+};
+
+Room.prototype.addTerminalToFillerArea = function() {
+  const fillerPos = this.memory.position.creep.filler;
+  const getNearPathPos = (pos) => Array.from(pos.getAllAdjacentPositions()).find((p) => p.inPath() && !p.isEqualTo(fillerPos));
```

Comment:
needs to use RoomPosition.prototype.validPosition or otherwise incorporate p.inPositions() to avoid putting the terminal filler on top of an existing placement.

Reply:
ok, let's make it as separate PR

Reply:
as-is, this code is going to cause traffic jams any time the towerfiller wants to path to a spot permanently occupied by another creep.

Reply:
Yeah, I agree, but if I change it - it will take another month to merge. On the other hand - quick one-liner PR will be merged quickly. Especially with reviewer approves. Moreover, this change merged as-is won't change layouts of existing rooms until `config.layout.version` bump from your layout changes, as we discussed in slack